### PR TITLE
Retry add member policy to resolve data collector token issue

### PR DESCRIPTION
**Pull request by jayvikramsharma1**
### :nut_and_bolt: Description: What code changed, and why?

sometime data-collector api fails and cause deployment to failed with automate, chef-server as critical state, actual root cause was default data collector token was not added in member policy, this situation may occur when authn service is up but authz service not available at that time, so it is not able to assign ingest policy.

In order to resolve above, we have added a retry mechanism with some sleep interval to add data-collector token to policy member.

### :+1: Definition of Done

Dev Done

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

**Review by punitmundra — CHANGES_REQUESTED** [review]
check the alignment

File: components/authn-service/server/server.go [hunk 1]
```diff
@@ -193,16 +194,26 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
+			})
+			if err != nil {
+				s.logger.Warn(errors.Wrap(err, "there was an error granting the legacy data collector token ingest access").Error())
+				s.logger.Warn(fmt.Sprintf("please manually add token with ID %q to the policy with ID %q", tokenID, IngestPolicyID))
+			} else {
+				s.logger.Warn("exiting for loop")
+				break
+			}
+			s.logger.Warn(fmt.Sprintf(" Adding the sleep for %d seconds ", i))
+			time.Sleep(time.Duration(i) * time.Second)
```
Comment (kalroy) [inline components/authn-service/server/server.go:214]: Having a variable sleep time may not be a good idea. The last iteration will have 1 second delay.

File: components/authn-service/server/server.go [hunk 2]
```diff
@@ -193,16 +194,26 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
+				Members: []string{fmt.Sprintf("token:%s", tokenID)},
+			})
+			if err != nil {
+				s.logger.Warn(errors.Wrap(err, "there was an error granting the legacy data collector token ingest access").Error())
+				s.logger.Warn(fmt.Sprintf("please manually add token with ID %q to the policy with ID %q", tokenID, IngestPolicyID))
+			} else {
+				s.logger.Warn("exiting for loop")
+				break
+			}
+			s.logger.Warn(fmt.Sprintf(" Adding the sleep for %d seconds ", i))
```
Comment (kalroy) [inline components/authn-service/server/server.go:213]: Should be debug rather warn

File: components/authn-service/server/server.go [hunk 3]
```diff
@@ -193,16 +194,26 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
+		for i := 4; i > 0; i-- {
+			_, err := policiesClient.AddPolicyMembers(ctx, &authz.AddPolicyMembersReq{
+				Id:      IngestPolicyID,
+				Members: []string{fmt.Sprintf("token:%s", tokenID)},
+			})
+			if err != nil {
+				s.logger.Warn(errors.Wrap(err, "there was an error granting the legacy data collector token ingest access").Error())
+				s.logger.Warn(fmt.Sprintf("please manually add token with ID %q to the policy with ID %q", tokenID, IngestPolicyID))
+			} else {
+				s.logger.Warn("exiting for loop")
```
Comment (kalroy) [inline components/authn-service/server/server.go:210]: Should be debug

File: components/authn-service/server/server.go [hunk 4]
```diff
@@ -193,16 +194,26 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
+			Adding the sleep, is just the tempory solution, this might work
+		*/
+		for i := 4; i > 0; i-- {
+			_, err := policiesClient.AddPolicyMembers(ctx, &authz.AddPolicyMembersReq{
+				Id:      IngestPolicyID,
+				Members: []string{fmt.Sprintf("token:%s", tokenID)},
+			})
+			if err != nil {
+				s.logger.Warn(errors.Wrap(err, "there was an error granting the legacy data collector token ingest access").Error())
+				s.logger.Warn(fmt.Sprintf("please manually add token with ID %q to the policy with ID %q", tokenID, IngestPolicyID))
```
Comment (kalroy) [inline components/authn-service/server/server.go:208]: This is not valid here. Should be logged when retry fails

**Review by kalroy — CHANGES_REQUESTED** [review]

File: components/authn-service/server/server.go (hunk 1 above)
Comment (vivek-yadav) [reply components/authn-service/server/server.go:214]: Have we tested that this much delay is enough always?

Comment (punitmundra) [reply components/authn-service/server/server.go:214]: yes..i have tested that.

**Comment by jayvikramsharma1** [comment]
@kalroy @vivek-yadav all review comments were resolved, please have a review again

**Review by vivek-yadav — APPROVED** [review]

**Comment by jayvikramsharma1** [comment]
@kalroy @vivek-yadav please have a review

File: components/authn-service/server/server.go [hunk 5]
```diff
@@ -193,16 +194,31 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
-			Members: []string{fmt.Sprintf("token:%s", tokenID)},
-		})
-		if err != nil {
+		/*
+			Below code is only execute when you add token in config at the time of deployment
+			Adding the sleep, is just the tempory solution, this might work
+		*/
+		var policyAssigned bool = false
+		for i := 4; i > 0; i-- {
+			_, err := policiesClient.AddPolicyMembers(ctx, &authz.AddPolicyMembersReq{
```
Comment (punitmundra) [inline components/authn-service/server/server.go:203]: err is the scope inside the for loop

File: components/authn-service/server/server.go [hunk 6]
```diff
@@ -193,16 +194,31 @@ func newServer(ctx context.Context, c Config) (*Server, error) {
...
-		_, err := policiesClient.AddPolicyMembers(ctx, &authz.AddPolicyMembersReq{
-			Id:      IngestPolicyID,
-			Members: []string{fmt.Sprintf("token:%s", tokenID)},
-		})
-		if err != nil {
+		/*
+			Below code is only execute when you add token in config at the time of deployment
+			Adding the sleep, is just the tempory solution, this might work
+		*/
+		var policyAssigned bool = false
```
Comment (punitmundra) [inline components/authn-service/server/server.go:201]: var err error ...will fix scope

**Review by punitmundra — CHANGES_REQUESTED** [review]

**Review by punitmundra — APPROVED** [review]

**Pull request merged**